device_service.go

```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"errors"
+	"log"
+)
+
+type Device struct {
+	Id       int     `json:"id"`
+	Name     string  `json:"name"`
+	Interval int     `json:"interval"`
+	Value    float32 `json:"value"`
+}
+
+type deviceDAO interface {
+	Save(device Device) (Device, error)
+}
+
+type deviceService struct {
+	dao deviceDAO
+}
+
+func (s *deviceService) createDevice(device Device) (Device, error) {
```

Comment:
How do you feel about changing the signature into `func (s deviceService) createDevice(device Device) (*Device, error)`? Do you know why I am asking about it?

Reply:
I prefer to stay with pointer receiver to prevent of creating copy of deviceDAO. For now it wouldn't be a problem but when we switch to real database copying connections will be bad idea.

Reply:
1. Neither pointer nor value receiver for `func (s *deviceService) createDevice` will prevent from copying `dao`. `var s *service` or `func t(s *service)` is very different from `func (s *service)` or manual copying of struct pointers.
2. Having singleton for `deviceDAO` would prevent from copying it
3. I was asking about value receiver because value receiver means that the struct is immutable (at least this method cannot change struct's content). This service is used concurrently (goroutine for each HTTP request) and therefore you don't want it to be mutable - https://play.golang.org/p/rm7nA1qRCEC
4. In case of services and DAOs in layered architecture I think problem of copying services/daos/connections is very rare - most of the times you have a factory of some kind or you are using CDI and it's set up only one time.

I know there was nothing dangerous in this code - I just wanted to discuss the receivers ;)